File: models/interfaceV1.py
from pydantic import BaseModel
from typing import List, Optional, Dict


class Controller(BaseModel):
    name: str
    type: str


class Resource(BaseModel):
    name: str
    path: List[str]


class Agent(BaseModel):
    child_exec: str
    child_args: List[str]


class Task(BaseModel):
    name: str
    entry: str
    doc: Optional[str | List[str]] = None
    option: Optional[List[str]] = None
    pipeline_override: Optional[Dict[str, dict]] = None
    repeatable: Optional[bool] = None
    repeat_count: Optional[int] = None


class OptionCase(BaseModel):
    name: str
    pipeline_override: Dict[str, dict]


class Option(BaseModel):
    default_case: Optional[str] = None
    cases: List[OptionCase]


class InterfaceModel(BaseModel):
    name: str
    url: str
    mirrorchyan_rid: Optional[str] = None
    mirrorchyan_multiplatform: Optional[bool] = None
    controller: List[Controller]
    resource: List[Resource]
    agent: Optional[Agent] = None
    task: List[Task]
    option: Dict[str, Option]
    version: str

class DeviceModel(BaseModel):
    name: str
    adb_path: str
    address: str
    screencap_methods: int
    input_methods: int
    config: dict

class PostTaskModel(BaseModel):
    tasklist: List[Dict[str, str]]

File: front/src/types/interfaceV1.ts
interface Controller {
  name: string
  type: string
}

interface Resource {
  name: string
  path: string[]
}

interface Agent {
  child_exec: string
  child_args: string[]
}

export interface Task {
  name: string
  entry: string
  doc?: string | string[]
  option?: string[]
  pipeline_override?: Record<string, object>
  repeatable?: boolean
  repeat_count?: number
}

export interface OptionCase {
  name: string
  pipeline_override: Record<string, object>
}

export interface Option {
  default_case?: string
  cases: OptionCase[]
}

export interface InterfaceModel {
  name: string
  url: string
  mirrorchyan_rid?: string
  mirrorchyan_multiplatform?: boolean
  controller: Controller[]
  resource: Resource[]
  agent?: Agent
  task: Task[]
  option: Record<string, Option>
  version: string
}
